storage/fanout.go

```diff
@@ -190,66 +215,80 @@ func (f *fanoutAppender) Rollback() (err error) {
 type mergeGenericQuerier struct {
 	mergeFunc genericSeriesMergeFunc
 
-	primaryQuerier genericQuerier
+	primaryQuerier int
 	queriers       []genericQuerier
 	failedQueriers map[genericQuerier]struct{}
 	setQuerierMap  map[genericSeriesSet]genericQuerier
 }
 
-// NewMergeQuerier returns a new Querier that merges results of chkQuerierSeries queriers.
-// NewMergeQuerier will return NoopQuerier if no queriers are passed to it
-// and will filter NoopQueriers from its arguments, in order to reduce overhead
-// when only one querier is passed.
-// The difference between primary and secondary is as follows: f the primaryQuerier returns an error, query fails.
-// For secondaries it just return warnings.
-func NewMergeQuerier(primaryQuerier Querier, queriers []Querier, mergeFunc VerticalSeriesMergeFunc) Querier {
+// NewMergeQuerier returns a new Querier that merges results of given slice of queriers.
+// NewMergeQuerier will return NoopQuerier if no queriers or noopQueriers are passed.
+// Function allows to optionally mark certain querier in the slice to be primary one. The difference between primary and
+// rest is as follows: if the primary querier returns an error, query fails. For secondaries it just return warnings.
+// Argument 'primaryQuerier' set to -1 means no primary querier.
+
+// In case of overlaps in the returned data by all queriers, mergeFunc will be used.
+func NewMergeQuerier(queriers []Querier, primaryQuerier int, mergeFunc VerticalSeriesMergeFunc) Querier {
 	filtered := make([]genericQuerier, 0, len(queriers))
-	for _, querier := range queriers {
+	for i, querier := range queriers {
 		if _, ok := querier.(noopQuerier); !ok && querier != nil {
 			filtered = append(filtered, newGenericQuerierFrom(querier))
+			continue
+		}
+		if i == primaryQuerier {
+			// Primary was filtered, no primary anymore.
```

Comment:
How can this happen?

Reply:
Well any consumer of this function can do this anytime, but just specifying `MergeQuerier([]Querier{NoopQuerier()}, 0, ...)` no?

Reply:
Do we have any of those in Prometheus, or are you keeping this general for Thanos?

Reply:
We don't have at the current moment, but what if write code that does it? We don't use this at all for Thanos.

Reply:
We can worry about it then, as-is we always have a primary querier and it's always present.

Reply:
The primaryQuerier argument to mergeGenericQuerier seemed unneeded now, and we're not handling it correctly now anyway.

Reply:
This still doesn't look correct.

Reply:
looking

Reply:
Hm, what do you mean?

PrimaryQuerier argument actually tells us if it should be error or warning e.g:

```
	           if q.primaryQuerier == 0 {
				return nil, w, err
			}
			// Nil will be converted to noop in adapter.
			return nil, append(w, err), nil
```

and

```
                if r.selectError != nil {
			q.failedQueriers[q.queriers[r.i]] = struct{}{}
			if r.i != q.primaryQuerier {
				// If the error source isn't the primary querier, return the error as a warning and continue.
				warnings = append(warnings, r.selectError)
				continue
			}
			priErr = r.selectError
			continue
		}
```
So it's used, and to me it looks correctly.

Reply:
@brian-brazil

Reply:
What's broken it's that the first if `len(queries) == 1` is not needed.

Reply:
Removed.

Reply:
This still doesn't look right (warnigns vs errors), and even if it is it's going to be fragile. I think passing the primary querier as an index rather than the actual object is the fundamental issue, so would suggest going back to having a primary querier and a list of secondaries.

Reply:
Hm tend to agree with this, but the previous logic was extremely surprising as we expected caller to fill all queries + primary additionally.

I think I would be fine with primary and secondaries as different args, I think that would  work for both of all (: